### PR TITLE
Mrc 5163 - Testing

### PR DESCRIPTION
**Pull request by weshinsley**

### CODE REVIEW

File: hipercow-api/Controllers/ClustersController.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    /// </summary>
+    [ApiController]
+    [Route("api/v1/[controller]")]
+    public class ClustersController : ControllerBase
+    {
+        /// <summary>
+        /// An object implementing IClusterInfoQuery, responsible for returning
+        /// the list of clusters, or information on a specific cluster.
+        /// </summary>
+        private IClusterInfoQuery clusterInfoQuery = new ClusterInfoQuery();
```
Comment (absternator) [inline hipercow-api/Controllers/ClustersController.cs:21]: should proabbly use dependency injection here... in program.cs `builder.Services.AddSingleton<IClusterInfoQuery, ClusterInfoQuery>();` .... and then here can inject into constructor.. also will allow you to mock and make testing easier
```
private IClusterInfoQuery clusterInfoQuery;

public ClustersController(IClusterInfoQuery clusterInfoQuery) 
{
this.clusterInfoQuery = clusterInfoQuery
}
```

File: hipercow-api/Controllers/ClustersController.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        /// <summary>
+        /// Used for testing. Allow injecting an IClusterInfoQuery object
+        /// returning defined results, rather than doing a genuine query.
+        /// </summary>
+        /// <param name="inject">A mock cluster info query.</param>
+        [NonAction]
+        public void MockClusterInfoQuery(IClusterInfoQuery inject)
```
Comment (absternator) [inline hipercow-api/Controllers/ClustersController.cs:29]: unsure if this is needed if you inject clusterInfoQuery and mock this class in tests

File: hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        /// <summary>
+        /// Test just the endpoint handling of /clusters/cluster
+        /// by providing fake data.
+        /// </summary>
+        [Fact]
+        public void GetClusterinfo_Works()
+        {
+            ClustersController cc = new ClustersController();
+            FakeClusterInfoQuery fake = new FakeClusterInfoQuery();
```
Comment (absternator) [inline hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs:47]: General comment that also appolies here.. probably use actual mocks for these (using Moq) can control these a lot better and avoid having to create fake classes... you will be able to do it like 
```
Mock<IClusterQuery> mockClusterQuery = new() 
mockClusterQuery.Setup(x => x.GetClusterInfo(It.IsAny<String>())).Returns(new ClusterInfo())
```

**Review by absternator — COMMENTED** [review]
Looks solid and great work!!! Just left main comments regarding fundamentals with .net --> 1. dependency injecting
2. using mocks for testing

File: hipercow-api/Tools/HipercowScheduler.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        /// <summary>
+        /// Populate testData placeholder, and return this from the
+        /// NodesQuery call if testing.
+        /// </summary>
+        /// <param name="data">The PropertyRowSet of fake data.</param>
+        public void SetTestData(PropertyRowSet data)
+        {
+            this.testData = data;
+        }
```
Comment (EmmaLRussell) [inline hipercow-api/Tools/HipercowScheduler.cs:45]: You shouldn't need to do stuff like this if you use mocking

File: hipercow-api-unit-tests/stylecop.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿{
+  // ACTION REQUIRED: This file was automatically added to your project, but it
+  // will not take effect until additional steps are taken to enable it. See the
+  // following page for additional information:
+  //
+  // https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/EnableConfiguration.md
```
Comment (EmmaLRussell) [inline hipercow-api-unit-tests/stylecop.json:6]: Has this dire warning been followed? Do you need to add stylecopy to your gh action or does it get triggered elsewhere?

File: hipercow-api/Controllers/ClustersController.cs (hunk 2 above)
Comment (EmmaLRussell) [reply hipercow-api/Controllers/ClustersController.cs:29]: Yeah, ideally not. It's almost always possible to find ways to unit test without adding extra test-only functionality to the module under test (though it does require supporting DI etc in the overall design). I think the clusterInfoQuery should be injectable in the constructor?

**Review by EmmaLRussell — COMMENTED** [review]
Kudos on the 100% coverage!  As Anmol says, I think these unit tests could be a bit more unity, but you probably would want to make some integration tests as well - either testing the full class system starting from the controllers, or full e2e tests involving spinning up the API and actually hitting the endpoints - or a combination of the two. But that can wait for another branch!

**Comment by weshinsley** [comment]
OK - in the interest of sanity, I'll merge this one and then learn how to do the things above in smaller PRs which should be much more reviewable...

File: hipercow-api/Controllers/ClustersController.cs (hunk 1 above)
Comment (weshinsley) [reply hipercow-api/Controllers/ClustersController.cs:21]: Think I have done this correctly in #3

File: hipercow-api/Controllers/ClustersController.cs (hunk 2 above)
Comment (weshinsley) [reply hipercow-api/Controllers/ClustersController.cs:29]: Yep, that disappears with the injection...

File: hipercow-api-unit-tests/stylecop.json (hunk 5 above)
Comment (weshinsley) [reply hipercow-api-unit-tests/stylecop.json:6]: Ah - yes, this has been followed (without it, I think stylecop doesn't work in some way! - But I haven't removed it from the template file. I'll zap it (and also in the main project) in #3

File: hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        /// <summary>
+        /// Test that /clusters tells us about wpia-hn.
+        /// </summary>
+        [Fact]
+        public void GetClusterCall_Works()
+        {
+            ClustersController cc = new ClustersController();
+            List<string> clusters = cc.Get();
+            Assert.Single(clusters);
+            Assert.Equal("wpia-hn", clusters.First());
```
[the remark this answers is not in the document]
Comment (weshinsley) [reply hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs:24]: Yep - Assert.Equal(new List<string> { "wpia-hn" }, clusters);

File: hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs (hunk 3 above)
Comment (weshinsley) [reply hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs:47]: Think this is done now in #3

File: README.md [hunk 7]
```diff
@@ -1,2 +1,8 @@
 # hipercow-api
 
+<!-- badges: start -->
+[![Project Status: WIP � Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
+[![codecov.io](https://codecov.io/gh/mrc-ide/hipercow-api/graph/badge.svg?token=1Zeyvs0lfh)](https://codecov.io/gh/mrc-ide/hipercow-api)
+<!-- badges: end -->
+
+This is the .NET Core Web api to talk to our MS HPC Cluster.
```
[the remark this answers is not in the document]
Comment (weshinsley) [reply README.md:8]: There isn't really a repo for the cluster itself... the web portal is the closest - I'll improve this a bit later...

File: hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+{
+    using Hipercow_api.Controllers;
+    using Hipercow_api.Models;
+    using Hipercow_api.Tools;
+    using Microsoft.AspNetCore.Mvc;
+
+    /// <summary>
+    /// Test the /clusters endpoint.
+    /// </summary>
+    public class ClustersControllerUnitTest
```
[the remark this answers is not in the document]
Comment (weshinsley) [reply hipercow-api-unit-tests/Controllers/ClustersControllerUnitTest.cs:13]: I think I've got this better in #3 - maybe still a little fuzzy between integration/unit tests, but hopefully better...

**Pull request merged**